modules/dkan/dkan_harvest/dkan_harvest.module

```diff
@@ -223,41 +223,76 @@ function dkan_harvest_page_tab_access($node) {
-    // Could not cache the source. issue?
-    $message = t("Failed to fetch the source. Please make sure that the source
-is reachable and the filters are set up correctly.");
-    drupal_set_message($message, 'error', FALSE);
-    $output = '<em>Preview is not available for this source. Please check the above errors.</em>';
-  }
-  else {
-    foreach ($harvestCache->getSaved() as $id => $dataset_item) {
-      $rows[] = array($id, $dataset_item['title']);
-    }
+  if (!$cache_dir) {
+    // The cache directory does not exist.
+    // Perform a cache operation.
+    $batch_operations = array();
+    $batch_operations[] = array('dkan_harvest_cache_source', array($node));
+
+    // Setup batch process.
+    $batch = array(
+      'operations' => $batch_operations,
+      'title' => t('Caching source and preparing Harvest preview...'),
+      'init_message' => t('Processed 0%.'),
+      'progress_message' => t('Processed @percentage.'),
+      'error_message' => t('An error ocurred while caching the source.'),
+    );
+    batch_set($batch);
+    batch_process();
+
+  } else {
+    // The cache directory exists. Load the table.
+    $harvestCache = call_user_func($harvestSource->type->cache_callback, $harvestSource, NULL);
+
+    if ($harvestCache) {
+      foreach ($harvestCache->getSaved() as $id => $dataset_item) {
+        $rows[] = array($id, $dataset_item['title']);
+      }
 
-    $form_import = drupal_get_form('dkan_harvest_harvest_now_form', $node);
-    $caption = t('Harvesting from this source as it is currently configured will
-  import the datasets listed below into this data portal. Click "Edit" above to
-  change the source\'s configuration, click the "Harvest now" button at the
-  bottom of the page to run the harvest immediately, or simply continue working
-  and the harvest will be performed on schedule.');
-    $header = array(t('Identifier'), t('Title'));
-    $output .= theme('table', array('rows' => $rows, 'header' => $header, 'caption' => $caption));
-    $output .= drupal_render($form_import);
+      $form_import = drupal_get_form('dkan_harvest_harvest_now_form', $node);
+      $caption = t('Harvesting from this source as it is currently configured will
+        import the datasets listed below into this data portal. Click "Edit" above to
+        change the source\'s configuration, click the "Harvest now" button at the
+        bottom of the page to run the harvest immediately, or simply continue working
+        and the harvest will be performed on schedule.');
+      $output .= '<div class="datasets-count-preview"><span>' .
+        t('Datasets in source: ') . '</span>' . $harvestCache->getSavedCount() . '</div>';
+      $header = array(t('Identifier'), t('Title'));
+      $output .= theme('table', array('rows' => $rows, 'header' => $header, 'caption' => $caption));
+      $output .= drupal_render($form_import);
+    }
   }
 
   return $output;
 }
 
+/**
+ *
+ */
+function dkan_harvest_cache_source($node, &$context) {
```

Comment:
This seems very redundant with dkan_harvest_cache_sources() - want to consolidate the code.